### DISCUSSION TITLE
Good afternoon! I have a problem with the work of @njit in combining arrays

### DISCUSSION THREAD

**Discussion by IvanDenisovich228** [community/community#61659] [Programming Help]
### Select Topic Area

Question

### Body

I have a problem. A simple example of combining numba arrays shows that the union does not work with @njit and gives errors.

'import numpy as np
from numba import njit
arr1 = np.array([1, 2, 3])
arr2 = np.array([4, 5, 6])
arr3 = np.array([7, 8, 9])
print('arr1 :', arr1)
print()
print('arr2 :', arr2)
print()
print('arr3 :', arr3)'
without @njit

'def union_of_arrays():
    
    arr = np.array([arr1, arr2, arr3]).T
    
    return arr

arr = union_of_arrays()
print(arr)'
Everything works, now with @njit

'@njit
def union_of_arrays():
    
    arr = np.array([arr1, arr2, arr3]).T
    
    return arr

arr = union_of_arrays()
print(arr)`
throws an error:

'TypingError                               Traceback (most recent call last)
Cell In[4], line 8
      4     arr = np.array([arr1, arr2, arr3]).T
      6     return arr
----> 8 arr = union_of_arrays()
      9 print(arr)

File c:\Users\user\AppData\Local\Programs\Python\Python310\lib\site-packages\numba\core\dispatcher.py:468, in _DispatcherBase._compile_for_args(self, *args, **kws)
    464         msg = (f"{str(e).rstrip()} \n\nThis error may have been caused "
    465                f"by the following argument(s):\n{args_str}\n")
    466         e.patch_message(msg)
--> 468     error_rewrite(e, 'typing')
    469 except errors.UnsupportedError as e:
    470     # Something unsupported is present in the user code, add help info
    471     error_rewrite(e, 'unsupported_error')

File c:\Users\user\AppData\Local\Programs\Python\Python310\lib\site-packages\numba\core\dispatcher.py:409, in _DispatcherBase._compile_for_args..error_rewrite(e, issue_type)
    407     raise e
    408 else:
--> 409     raise e.with_traceback(None)

TypingError: Failed in nopython mode pipeline (step: nopython frontend)
No implementation of function Function() found for signature:
 
...

File "..\..\AppData\Local\Temp\ipykernel_8316\2415862080.py", line 4:'

How to fix the error so that the code would work?

The code is the same, but it doesn't work with @njit

**Comment by queenofcorgis** [comment]
Thanks for posting in the GitHub Community, @IvanDenisovich228 ! 

We’ve moved your post to our new Programming Help 🧑‍💻 category, which is more appropriate for this type of discussion.

Please review our [guidelines](https://github.com/orgs/community/discussions/51384) about the Programming Help category for more information.

**Comment by Beargruug** [comment]
The error you encountered with the @njit decorator is because Numba's just-in-time (JIT) compiler does not support the use of global variables inside the decorated function. In your example, arr1, arr2, and arr3 are global variables, and using them inside the @njit function causes the typing error.

To fix this issue, you can pass the arrays arr1, arr2, and arr3 as arguments to the union_of_arrays function. Here's how you can modify your code:

```
import numpy as np
from numba import njit

@njit
def union_of_arrays(arr1, arr2, arr3):
    arr = np.array([arr1, arr2, arr3]).T
    return arr

arr1 = np.array([1, 2, 3])
arr2 = np.array([4, 5, 6])
arr3 = np.array([7, 8, 9])

arr = union_of_arrays(arr1, arr2, arr3)
print(arr)
```

By passing the arrays as arguments, you make sure that the @njit function has all the required information during compilation, and this will resolve the typing error.
I hope this helps you!
